packages/core/src/libraries/quota.ts

```diff
@@ -156,16 +175,107 @@ export const createQuotaLibrary = (
-      await reportSubscriptionUpdates(cloudConnection, key);
+    if (this.shouldReportSubscriptionUpdates(planId, isEnterprisePlan, key)) {
+      await reportSubscriptionUpdates(this.cloudConnection, key);
     }
   };
 
-  return {
-    guardTenantUsageByKey,
-    guardEntityScopesUsage,
-    reportSubscriptionUpdatesUsage,
-  };
-};
+  public async getTenantUsage(): Promise<{ usage: SelfComputedTenantUsage }> {
+    const rawUsage = await this.getRawTenantUsage();
+
+    const connectors = await this.connectorLibrary.getLogtoConnectors();
+    const socialConnectors = connectors.filter(
+      (connector) => connector.type === ConnectorType.Social
+    );
+
+    const unparsedUsage: SelfComputedTenantUsage = {
+      applicationsLimit: rawUsage.applicationsLimit,
+      thirdPartyApplicationsLimit: rawUsage.thirdPartyApplicationsLimit,
+      scopesPerResourceLimit: rawUsage.scopesPerResourceLimit, // Max scopes per resource
+      userRolesLimit: rawUsage.userRolesLimit,
+      machineToMachineRolesLimit: rawUsage.machineToMachineRolesLimit,
+      scopesPerRoleLimit: rawUsage.scopesPerRoleLimit, // Max scopes per role
+      hooksLimit: rawUsage.hooksLimit,
+      customJwtEnabled: rawUsage.customJwtEnabled,
+      bringYourUiEnabled: rawUsage.bringYourUiEnabled,
+      /** Add-on quotas start */
+      machineToMachineLimit: rawUsage.machineToMachineLimit,
+      resourcesLimit: rawUsage.resourcesLimit,
+      enterpriseSsoLimit: rawUsage.enterpriseSsoLimit,
+      mfaEnabled: rawUsage.mfaEnabled,
+      /** Enterprise only add-on quotas */
+      idpInitiatedSsoEnabled: rawUsage.idpInitiatedSsoEnabled,
+      samlApplicationsLimit: rawUsage.samlApplicationsLimit,
+      socialConnectorsLimit: socialConnectors.length,
+      organizationsLimit: rawUsage.organizationsLimit,
+      /**
+       * We can not calculate the quota usage since there is no related DB configuration for such feature.
+       * Whether the feature is enabled depends on the `quota` defined for each plan/SKU.
+       * If we mark this value as always `true`, it could block the subscription downgrade (to free plan) since in free plan we do not allow impersonation feature.
+       */
+      subjectTokenEnabled: false,
+    };
+
+    return { usage: selfComputedSubscriptionUsageGuard.parse(unparsedUsage) };
+  }
+
+  public async getScopesForRolesTenantUsage() {
+    const records = await this.pool.any<{ roleId: string; count: string }>(
+      this.sqlBuilder.countScopesForRoles()
+    );
+
+    return Object.fromEntries(records.map(({ roleId, count }) => [roleId, Number(count)]));
+  }
+
+  public async getScopesForResourcesTenantUsage() {
+    const records = await this.pool.any<{ resourceId: string; count: string }>(
+      this.sqlBuilder.countScopesForResources()
+    );
+
+    return Object.fromEntries(records.map(({ resourceId, count }) => [resourceId, Number(count)]));
+  }
+
+  private async getRawTenantUsage() {
```

Comment:
Should we put all these query builders under the `query` directory instead of `library`?